### PR TITLE
Initial implementation of PyTest Framework

### PR DESCRIPTION
**Pull request by shots47s**
---
name: Added Unit Testing through the PyTest Framework
about: added tests directory
title: 'Feature implementation: Pytesting'
labels: enhancement
assignees: ''

---

## Related issues
None

## Checklist

- [x] **DO** Unit tests pass.
- [X] **DO** If new feature, created unit test.

## Purpose
Added a pytesting framework with some example Unit, Database and blueprint tests.

## Current behaviour
No testing

## New behaviour
pytest now runs a series of tests, and more can be created.
 
#### Does this introduce a major change?
- [ X] Yes
- [ ] No

## Implementation Detail
There is a new directory, /tests.  Check out conftest.py for the fixtures that create and teardown the test build, it is pretty cool.  Pytest rules!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,5 +10,5 @@ install:
 - pip install -r requirements.txt
 
 script:
-- pycodestyle --max-line-length=99 app/*.py app/*/*.py
+- pycodestyle --max-line-length=99 -r *.py app/*.py
```
Comment (glatard) [inline .travis.yml]: Shouldn't Travis be running `pytest` here?

Comment (glatard) [reply .travis.yml]: You should also configure coveralls, to measure test coverage, and prevent merging PRs that lead to decrease in coverage. See example here: https://github.com/boutiques/boutiques/blob/master/.travis.yml#L53-57

Comment (shots47s) [reply .travis.yml]: Good suggestion, doing so now.

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+
+@pytest.fixture(scope='session')
+def app(request):
+    """
+    This is creates the mock app for testing, it uses the
+    TestingConfig in Config.py
+    """
+    app = create_app(config_settings=TestingConfig)
+    ctx = app.app_context()
```
[the remark this answers is not in the document]
Comment (shots47s) [reply tests/conftest.py]: hmm, ok, I will give it a try, not sure why that was happening. It was passing for me, did you update your pip installs?  There are some new modules that need to be added, I should have said that.

[the remark this answers is not in the document]
Comment (liamocn) [reply tests/conftest.py]: yes i updated dependencies. Strange it was passing for you but the travis builds were definitely failing too for the same reason I had.

File: .travis.yml (hunk 1 above)
Comment (shots47s) [reply .travis.yml]: Added as you can see below.

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+@pytest.fixture(scope='session')
+def app(request):
+    """
+    This is creates the mock app for testing, it uses the
+    TestingConfig in Config.py
+    """
+    app = create_app(config_settings=TestingConfig)
+    ctx = app.test_request_context()
```
Comment (shots47s) [inline tests/conftest.py]: Added the suggestion, after reading it is actually the right thing to do.  @liamocn give it another try.

**Pull request merged**